TropicalGeometry: fix tropical_variety_prime_singular

YueRen opened the pull request:
added function to make variable names singular compatible. fixes https://github.com/oscar-system/Oscar.jl/issues/5392

YueRen committed:
TropicalGeometry: fix tropical_variety_prime_singular

added function to make variable names singular compatible.
fixes https://github.com/oscar-system/Oscar.jl/issues/5392

fieker commented on src/TropicalGeometry/variety_prime.jl:
```diff
@@ -19,8 +19,23 @@ function tropical_variety_prime(I::MPolyIdeal, nu::TropicalSemiringMap; weighted
+    newVarNames = [ "x" * lpad(string(i),l,'0') for i in 1:n ]
+    S,x = polynomial_ring(base_ring(R), newVarNames)
+    changeVarNames = hom(R,S,x)
+    return changeVarNames(I)
```
I'd rather return the map: in other situations you might need to reinterpret the Singular result and relate them back to the original ring.

YueRen replied:
Hm, true.  I'll change the code accordingly later.

YueRen replied:
The function is now takes a multivariate polynomial ring and returns the homomorphism.

fieker commented:
apart from this working, this is still a cludge

benlorenz commented:
Can you add the example from the issue as a testcase?

YueRen committed:
TropicalGeometry: make renaming function return hom instead of ideal

YueRen committed:
TropicalGeometry: remove redundant empty line

YueRen committed:
TropicalGeometry: tropical_variety, added test

https://github.com/oscar-system/Oscar.jl/issues/5392

YueRen commented:
> Can you add the example from the issue as a testcase?

Is done!

lgoettgens commented on src/TropicalGeometry/variety_prime.jl:
```diff
@@ -19,8 +19,23 @@ function tropical_variety_prime(I::MPolyIdeal, nu::TropicalSemiringMap; weighted
+function hom_rename_variables_singular_compatible(S::MPolyRing)
+    n = nvars(S)
+    l = ndigits(n)
+    newVarNames = [ "x" * lpad(string(i),l,'0') for i in 1:n ]
```
```suggestion
    newVarNames = [Symbol(:x, lpad(string(i), l, '0')) for i in 1:n]
```
This should avoid some allocations

YueRen committed:
Update src/TropicalGeometry/variety_prime.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

YueRen committed:
TropicalGeometry: replaced dirty hack with a less dirty hack

removed the conversion of Oscar ideal to Singular via string

YueRen commented:
I've now replaced the original dirty hack with the suggestion by @fingolfin.  No conversion of an OSCAR ideal to a Singular string necessary anymore, but the need to call the Singular interpreter and the conversion of a Singular fan string to an OSCAR polyhedral complex remain.

fingolfin commented on src/TropicalGeometry/variety_prime.jl:
```diff
@@ -40,6 +43,8 @@ end
 # p-adic valuation
 function tropical_variety_prime_singular(I::MPolyIdeal, nu::TropicalSemiringMap{QQField,ZZRingElem,<:Union{typeof(min),typeof(max)}}; weighted_polyhedral_complex_only::Bool=false)
+    phi = hom_rename_variables_singular_compatible(base_ring(I))
+    I = phi(I)
```
Can this be removed again now?
```suggestion
```

fingolfin commented on src/TropicalGeometry/variety_prime.jl:
```diff
@@ -40,6 +43,8 @@ end
+    phi = hom_rename_variables_singular_compatible(base_ring(I))
+    I = phi(I)
     R = base_ring(I)
     singularCommand = join(["ring r=0,("*join(string.(symbols(R)),",")*"),dp;",
```
Shouldn't the code here be treated similarly?

YueRen replied:
Oops, forgot that there was more than one of them.  Fixing it now.

fingolfin commented on src/TropicalGeometry/variety_prime.jl:
```diff
@@ -19,16 +19,19 @@ function tropical_variety_prime(I::MPolyIdeal, nu::TropicalSemiringMap; weighted
+    if Singular.libSingular.have_error()
+        error(Singular.libSingular.get_and_clear_error())
+    end
+    TropIString = Singular.convert_return(return_value, ring)
```
I think this should also work, as the first "libname" argument to `Singular.low_level_caller` is only used to ensure that library is loaded (which is the case here anyway, so it basically does nothing), and then the lookup of the second argument (funcname) ignores the library name.
```suggestion
    sI = Oscar.singular_generators(I)
    TropIString = Singular.low_level_caller("tropical", "tropicalVariety_as_string", sI)
```

However, I did not test it. But if it works, it is clearly preferable :-)

YueRen replied:
Seems to be working.

YueRen replied:
@fingolfin do you happen to know how to call the following Singular function using `Singular.low_level_caller` for two inputs, the ideal `sI` and an integer `2`:

```
Singular.call_interpreter("proc tropicalVariety_as_string(ideal I, list #) { if(size(#)==0) { return(string(tropicalVariety(I))); }; return(string(tropicalVariety(I,#[1]))); }")
```

For the old version using `Singular.libSingular.call_singular_library_procedure` the following arguments worked:
```
arguments = Any[ Singular.prepare_argument(sI)[1], Singular.prepare_argument(Int(p))[1] ]
```

YueRen replied:
Nvm, I take back my comment that `arguments` above worked.  It works for a dummy function:

```
julia> Singular.call_interpreter("proc foo2(ideal I, list #) { print(#[1]); }") # function to call
4-element Vector{Any}:
 false
      ""
      ""
      ""

julia> I = grassmann_pluecker_ideal(2, 5); # argument 1

julia> p = ZZ(3); # argument 2

julia> sI = Oscar.singular_generators(I);

julia> sR = base_ring(sI)
Singular polynomial ring (QQ),(x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10),(dp(10),C)

julia> sp = Int(p)
3

julia> arguments = Any[ Singular.prepare_argument(sI)[1], Singular.prepare_argument(sp)[1] ]
2-element Vector{Any}:
 Any[276, Ptr{Nothing}(0x000000001176b2a0)]
 Any[420, Ptr{Nothing}(0x0000000000000003)]

julia> return_value = Singular.libSingular.call_singular_library_procedure("foo2", sR.ptr, arguments);
3

```

But somehow not for `tropicalVariety_as_string` which returns an empty string:

```
julia> Singular.libSingular.load_library("tropical.lib")
true

julia> Singular.call_interpreter("proc tropicalVariety_as_string(ideal I, list #) { if(size(#)==0) { return(string(tropicalVariety(I))); }; return(string(tropicalVariety(I,#[1]))); }") # function to call
4-element Vector{Any}:
 false
      ""
      ""
      ""

julia> I = grassmann_pluecker_ideal(2, 5); # argument 1

julia> nu = tropical_semiring_map(QQ, 3)
Map into Min tropical semiring encoding the 3-adic valuation on Rational field

julia> p = Oscar.uniformizer_ring(nu) # argument 2
3

julia> sI = Oscar.singular_generators(I);

julia> sR = base_ring(sI)
Singular polynomial ring (QQ),(x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10),(dp(10),C)

julia> sp = Int(p)
3

julia> arguments = Any[ Singular.prepare_argument(sI)[1], Singular.prepare_argument(sp)[1] ]
2-element Vector{Any}:
 Any[276, Ptr{Nothing}(0x0000000011753af8)]
 Any[420, Ptr{Nothing}(0x0000000000000003)]

julia> return_value = Singular.libSingular.call_singular_library_procedure("tropicalVariety_as_string", sR.ptr, arguments);

julia> return_value
3-element Vector{Any}:
 false
      Ptr{Nothing}(0x0000000011741750)
   511

julia> TropIString = Singular.convert_return(return_value, ring)
""

```

I will reinstate the old version for the second function, since at least it did work.

YueRen committed:
Update src/TropicalGeometry/variety_prime.jl

Co-authored-by: Max Horn <max@quendi.de>

YueRen committed:
chg: updating Singular call, reinstating `hom_rename_variables_singular_compatible`

YueRen committed:
TropicalGeometry: fixed test for tropical_variety

fingolfin commented on src/TropicalGeometry/variety_prime.jl:
```diff
@@ -38,8 +35,23 @@ function tropical_variety_prime_singular(I::MPolyIdeal, nu::TropicalSemiringMap{
+    R,x = polynomial_ring(base_ring(S), newVarNames)
+    return hom(S,R,x)
+end
+
```
```suggestion
```

YueRen committed:
TropicalGeometry: cleanup tropical varieties with p-adic valuation

YueRen committed:
TropicalGeometry: more cleanup

YueRen commented:
@fingolfin Everything should be using `Singular.low_level_caller` now.

fingolfin commented:
Thanks @YueRen this looks good. Unfortunately it seems to cause a crash :-( in CI, in the doctests:
```
page: src/TropicalGeometry/variety.jl:74-96
  0.132471 seconds (35.74 k allocations: 2.843 MiB, 56.50% compilation time)
page: src/TropicalGeometry/variety.jl:167-220
/home/runner/work/_temp/d5e2334b-f0b4-40c5-a276-f9262c06c6fc.sh: line 7:  2263 Segmentation fault      (core dumped)
```

fingolfin committed:
cleanup

fingolfin committed:
Adjust for Singular.jl changes

fingolfin commented on src/Oscar.jl:
```diff
@@ -178,6 +178,19 @@ function __init__()
+  # see https://github.com/oscar-system/Oscar.jl/issues/5392
+  # see TropicalGeometry/variety_prime.jl
+  Singular.libSingular.load_library("tropical.lib")
+  Singular.call_interpreter("""
+    proc tropicalVariety_as_string(ideal I, list #) {
+      if(size(#)==0) {
+        return(string(tropicalVariety(I)));
+      };
+      return(string(tropicalVariety(I,number(#[1]))));
+    }
+    """)
```
@YueRen I took the liberty of reformatting this for readability :-)

(I had to edit this PR anyway to adjust for the changed `Singular.low_level_caller` arguments introduced in the latest Singular.jl

fingolfin commented:
The crash is in a doctest in `src/TropicalGeometry/variety.jl`. I reproduced it locally and narrowed it down to this example:
```julia-repl
julia> K,t = rational_function_field(GF(101),:t);

julia> R,(x,y,z) = K["x","y","z"];

julia> nu_2 = tropical_semiring_map(QQ,2)
Map into Min tropical semiring encoding the 2-adic valuation on Rational field

julia> f1 = 8*x^2 + x*y + x*z + x + 8*y^2 + y*z + y + 8*z^2 + z + 8;

julia> f2 = x + 2;

julia> I = ideal([f1,f2]);

julia> TropI_2 = tropical_variety(I,nu_2; skip_saturation=true, skip_decomposition=true)

[76002] signal 11 (2): Segmentation fault: 11
in expression starting at REPL[13]:1
unknown function (ip: 0x0) at (unknown file)
Allocations: 99607851 (Pool: 99573602; Big: 34249); GC: 34
/Users/mhorn/.local/bin/j12: line 2: 76002 Segmentation fault: 11  julia +1.12 --proj=p/1.12 "$@"
```

Debugger backtrace:
```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x0000000000000000
    frame #1: 0x0000000333208bd8 libSingular-4.4.1.dylib`tropicalStrategy::tropicalStrategy(sip_sideal*, snumber*, ip_sring*) + 1528
    frame #2: 0x000000033321c3d0 libSingular-4.4.1.dylib`tropicalVariety(sleftv*, sleftv*) + 1920
    frame #3: 0x0000000333166ab0 libSingular-4.4.1.dylib`iiMake_proc(idrec*, sip_package*, sleftv*) + 320
    frame #4: 0x000000033312724c libSingular-4.4.1.dylib`jjPROC(sleftv*, sleftv*, sleftv*) + 220
    frame #5: 0x000000033312edf0 libSingular-4.4.1.dylib`iiExprArith2TabIntern(sleftv*, sleftv*, int, sleftv*, int, sValCmd2 const*, int, int, sConvertTypes const*) + 2096
    frame #6: 0x000000033312f904 libSingular-4.4.1.dylib`iiExprArith2(sleftv*, sleftv*, int, sleftv*, int) + 612
    frame #7: 0x0000000333153000 libSingular-4.4.1.dylib`jjKLAMMER_PL(sleftv*, sleftv*) + 176
    frame #8: 0x00000003331323c8 libSingular-4.4.1.dylib`iiExprArithM(sleftv*, sleftv*, int) + 952
    frame #9: 0x000000033312581c libSingular-4.4.1.dylib`yyparse() + 9596
    frame #10: 0x0000000333166420 libSingular-4.4.1.dylib`iiAllStart(procinfo*, char const*, feBufferTypes, int) + 272
    frame #11: 0x00000003331667e8 libSingular-4.4.1.dylib`iiPStart(idrec*, sleftv*) + 440
    frame #12: 0x0000000333166b70 libSingular-4.4.1.dylib`iiMake_proc(idrec*, sip_package*, sleftv*) + 512
    frame #13: 0x000000033312724c libSingular-4.4.1.dylib`jjPROC(sleftv*, sleftv*, sleftv*) + 220
    frame #14: 0x000000033312edf0 libSingular-4.4.1.dylib`iiExprArith2TabIntern(sleftv*, sleftv*, int, sleftv*, int, sValCmd2 const*, int, int, sConvertTypes const*) + 2096
    frame #15: 0x000000033312f904 libSingular-4.4.1.dylib`iiExprArith2(sleftv*, sleftv*, int, sleftv*, int) + 612
    frame #16: 0x0000000333153000 libSingular-4.4.1.dylib`jjKLAMMER_PL(sleftv*, sleftv*) + 176
    frame #17: 0x00000003331323c8 libSingular-4.4.1.dylib`iiExprArithM(sleftv*, sleftv*, int) + 952
    frame #18: 0x000000033312581c libSingular-4.4.1.dylib`yyparse() + 9596
    frame #19: 0x0000000333166420 libSingular-4.4.1.dylib`iiAllStart(procinfo*, char const*, feBufferTypes, int) + 272
    frame #20: 0x00000003331667e8 libSingular-4.4.1.dylib`iiPStart(idrec*, sleftv*) + 440
    frame #21: 0x0000000333166b70 libSingular-4.4.1.dylib`iiMake_proc(idrec*, sip_package*, sleftv*) + 512
    frame #22: 0x00000003331679cc libSingular-4.4.1.dylib`ii_CallLibProcM(char const*, void**, int*, ip_sring*, int&) + 492
    frame #23: 0x0000000333e202c8 libsingular_julia.dylib`call_singular_library_procedure(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, ip_sring*, jlcxx::ArrayRef<_jl_value_t*, 1>) + 280
    frame #24: 0x0000000333e249ac libsingular_julia.dylib`std::__1::__function::__func<_jl_value_t* (*)(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, ip_sring*, jlcxx::ArrayRef<_jl_value_t*, 1>), std::__1::allocator<_jl_value_t* (*)(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, ip_sring*, jlcxx::ArrayRef<_jl_value_t*, 1>)>, _jl_value_t* (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, ip_sring*, jlcxx::ArrayRef<_jl_value_t*, 1>)>::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&&, ip_sring*&&, jlcxx::ArrayRef<_jl_value_t*, 1>&&) + 60
    frame #25: 0x0000000333e2481c libsingular_julia.dylib`jlcxx::detail::CallFunctor<_jl_value_t*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, ip_sring*, jlcxx::ArrayRef<_jl_value_t*, 1>>::apply(void const*, jlcxx::WrappedCppPtr, jlcxx::WrappedCppPtr, jl_array_t*) + 92
    frame #26: 0x00000003d6eb4094
    frame #27: 0x00000001016416a0 libjulia-internal.1.12.1.dylib`eval_stmt_value(stmt=<unavailable>, s=<unavailable>) at interpreter.c:194:23 [opt]
    frame #28: 0x0000000101640320 libjulia-internal.1.12.1.dylib`eval_body(stmts=<unavailable>, s=0x000000016ef84560, ip=<unavailable>, toplevel=1) at interpreter.c:689:21 [opt]
    frame #29: 0x000000010164097c libjulia-internal.1.12.1.dylib`jl_interpret_toplevel_thunk(m=0x000000011c141fb0, src=0x0000000364ce6410) at interpreter.c:898:21 [opt]
    frame #30: 0x000000010165b338 libjulia-internal.1.12.1.dylib`jl_toplevel_eval_flex(m=0x000000011c141fb0, e=<unavailable>, fast=1, expanded=1, toplevel_filename=<unavailable>, toplevel_lineno=0x000000033e40f3b0) at toplevel.c:1035:18 [opt]
    frame #31: 0x0000000113e8748c u0gqU_vMi9E.dylib`japi1___repl_entry_eval_expanded_with_loc_19063.1 at REPL.jl:301
    frame #32: 0x000000010162d0ac libjulia-internal.1.12.1.dylib`jl_apply(args=<unavailable>, nargs=5) at julia.h:2391:12 [opt] [inlined]
    frame #33: 0x000000010162d09c libjulia-internal.1.12.1.dylib`jl_f_invokelatest(F=<unavailable>, args=<unavailable>, nargs=5) at builtins.c:881:23 [opt]
    frame #34: 0x000000011171c0f0
(lldb)
```

Perhaps @hannes14 or @YueRen can make something out of this.

fingolfin commented:
I would hope that @hannes14 @ederc or @YueRen might be able to look at this crash here?

(It'd also help if someone resolved the merge conflict that crept in in the meantime)

benlorenz commented on src/TropicalGeometry/variety_prime.jl:
```diff
@@ -39,15 +33,9 @@ function tropical_variety_prime_singular(I::MPolyIdeal, nu::TropicalSemiringMap{
-    TropIString = Singular.lookup_library_symbol("Top", "TropIString")
+function tropical_variety_prime_singular(I::MPolyIdeal, nu::TropicalSemiringMap{QQField,ZZRingElem}; weighted_polyhedral_complex_only::Bool=false)
+    sI = Oscar.singular_generators(I)
+    TropIString = Singular.low_level_caller("tropical", "tropicalVariety_as_string", (sI, Int(uniformizer_ring(nu))))
```
```suggestion
    TropIString = Singular.low_level_caller("tropical", "tropicalVariety_as_string", (sI, Int(uniformizer_in_ring(nu))))
```

benlorenz committed:
Update src/TropicalGeometry/variety_prime.jl

fingolfin commented:
To resolves this, it might help to move the definition of the helper function to Singular.

Hans told me before Christmas that the crash is caused by the interpreter "not knowing" about something... either this helper function, or (more likely I think) the polynomial ring.... This might help someone to figure this out.

I think this kind of approach followed in this PR would be very important to have working, as it opens the door for easy access to a bunch of other Singular goodies without having to write C++ code, recompile JLLs etc.

HereAround commented on src/Oscar.jl:
```diff
@@ -184,6 +184,19 @@ function __init__()
+  # see TropicalGeometry/variety_prime.jl
+  Singular.libSingular.load_library("tropical.lib")
+  Singular.call_interpreter("""
+    proc tropicalVariety_as_string(ideal I, list #) {
```
@YueRen We just touched on this PR in triage, and @jankoboehm suggested the following way forward to debug the failures:

1. Could you please try to integrate this procedure directly into Singular and try again. Does it work in Singular? If so, can you please open a correspponding PR to Singular. (Maybe you already tried this?)
2. Try out in OSCAR.

(And if that does not fix the issue, we will need to brainstorm again.)

YueRen committed:
TropicalGeometry: variety_prime.jl, fixed bad rebase

YueRen committed:
TropicalGeometry: tropical_variety, fixed docstring

YueRen commented:
Quite embarrassing, but I noticed that the crash is caused by a problem in my docstring (weird combinations of input that shouldn't be).  If that is indeed the case, I'll try to add some guardrails to prevent users running into it accidentally.

YueRen committed:
TropicalGeometry: tropical_variety_prime_singular, restrict input ideal type

YueRen committed:
TropicalGeometry: tropical_variety, fix docstring

YueRen committed:
TropicalGeometry: gfan_fan_string_to_oscar_complex, fixed dehomogenization

YueRen committed:
TropicalGeometry: tropical_variety, fixed docstring

YueRen committed:
TropicalGeometry: gfan_fan_string_to_oscar_complex, fixed dehomogenization

YueRen committed:
TropicalGeometry: tropical_variety_prime_singular, fixed convention logic

YueRen committed:
TropicalGeometry: added some more assertions

YueRen committed:
TropicalGeometry: tropical_variety_prime_singular, relaxing signature

YueRen commented:
Okay, the PR is ready to be merged, maybe after a final brief review.  Here are the main changes:

- There was a problem in the `tropical_variety` documentation:  One example worked over $\mathbb{Q}(t)$ equipped with the $p$-adic valuation, which the underlying Singular code was not designed for and led to a crash.  The example is now fixed and I've put some restrictions of the input types for `tropical_variety_prime_singular` to prevent it from happening.
- Instead of passing the ideal as a string (which caused https://github.com/oscar-system/Oscar.jl/issues/5392) and running everything via `Singular.call_interpreter`, the ideal is now passed via `Oscar.singular_generators` and the function is called via `Singular.low_level_caller`.  To facilitate that, `Oscar.jl` now loads the Singular library `tropical.lib` and defines a new Singular function `tropicalVariety_as_string`.
- There was a bug in the logic of `gfan_fan_string_to_oscar_complex` that is now fixed.  The bug was in the part which takes a polyhedral fan in $\mathbb{R}^{n+1}$, intersects it with $\\{1\\}\times\mathbb{R}^n$, and makes it into a polyhedral complex in $\mathbb{R}^n$.  And the old code simply forgot to divide the rays of $\mathbb{R}^{n+1}$ that represent vertices in $\mathbb{R}^n$ by their first coordinate, leading to a wrong output of `tropical_variety`.

YueRen committed:
TropicalGeoemtry: tropical_variety_prime_singular, simplified signature
